File: app/src/main/java/com/baikas/sporthub6/models/TerrainAddress.java
package com.baikas.sporthub6.models;

import com.baikas.sporthub6.interfaces.ObjectWithId;

import java.io.Serializable;
import java.util.Map;
import java.util.Objects;

public class TerrainAddress implements ObjectWithId, Serializable  {

    private String id;
    private String addressTitle;
    private String address;
    private String sport;
    private long priority;
    private String creatorId;
    private double latitude;
    private double longitude;
    private long createdAtUTC;

    public TerrainAddress(String id) {
        this.id = id;
    }

    public TerrainAddress(String id, String addressTitle, String address, String sport,long priority, String creatorId, double latitude, double longitude, long createdAtUTC) {
        this.id = id;
        this.addressTitle = addressTitle;
        this.address = address;
        this.sport = sport;
        this.priority = priority;
        this.creatorId = creatorId;
        this.latitude = latitude;
        this.longitude = longitude;
        this.createdAtUTC = createdAtUTC;
    }

    public TerrainAddress(Map<String,Object> map) {
        this.id = (String)map.get("id");
        this.addressTitle = (String)map.get("addressTitle");
        this.address = (String)map.get("address");
        this.sport = (String)map.get("sport");

        if (map.get("priority") instanceof Integer){
            this.priority = ((Integer) map.get("priority")).longValue();
        }else
            this.priority = (Long) map.get("priority");


        this.creatorId = (String)map.get("creatorId");
        this.latitude = (Double)map.get("latitude");
        this.longitude = (Double)map.get("longitude");

        if (map.get("createdAtUTC") instanceof Integer){
            this.createdAtUTC = ((Integer) map.get("createdAtUTC")).longValue();
        }else
            this.createdAtUTC = (Long) map.get("createdAtUTC");

    }

    public TerrainAddress(TerrainAddress terrainAddress) {
        this.id = terrainAddress.getId();
        this.addressTitle = terrainAddress.getAddressTitle();
        this.address = terrainAddress.getAddress();
        this.sport = terrainAddress.getSport();
        this.priority = terrainAddress.getPriority();
        this.creatorId = terrainAddress.getCreatorId();
        this.latitude = terrainAddress.getLatitude();
        this.longitude = terrainAddress.getLongitude();

        this.createdAtUTC = terrainAddress.getCreatedAtUTC();
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        TerrainAddress that = (TerrainAddress) o;

        if (Double.compare(that.latitude, latitude) != 0) return false;
        if (Double.compare(that.longitude, longitude) != 0) return false;
        if (createdAtUTC != that.createdAtUTC) return false;
        if (!Objects.equals(id, that.id)) return false;
        if (!Objects.equals(addressTitle, that.addressTitle))
            return false;
        if (!Objects.equals(address, that.address)) return false;
        if (!Objects.equals(sport, that.sport)) return false;
        if (!Objects.equals(priority, that.priority)) return false;
        return Objects.equals(creatorId, that.creatorId);
    }

    @Override
    public int hashCode() {
        int result;
        long temp;
        result = id != null ? id.hashCode() : 0;
        result = 31 * result + (addressTitle != null ? addressTitle.hashCode() : 0);
        result = 31 * result + (address != null ? address.hashCode() : 0);
        result = 31 * result + (sport != null ? sport.hashCode() : 0);
        result = 31 * result + (creatorId != null ? creatorId.hashCode() : 0);
        temp = Double.doubleToLongBits(latitude);
        result = 31 * result + (int) (temp ^ (temp >>> 32));
        temp = Double.doubleToLongBits(longitude);
        result = 31 * result + (int) (temp ^ (temp >>> 32));
        result = 31 * result + (int) (createdAtUTC ^ (createdAtUTC >>> 32));
        result = 31 * result + (int) (priority ^ (priority >>> 32));
        return result;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getAddressTitle() {
        return addressTitle;
    }

    public void setAddressTitle(String addressTitle) {
        this.addressTitle = addressTitle;
    }

    public String getAddress() {
        return address;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    public String getSport() {
        return sport;
    }

    public void setSport(String sport) {
        this.sport = sport;
    }

    public long getPriority() {
        return priority;
    }

    public void setPriority(long priority) {
        this.priority = priority;
    }

    public String getCreatorId() {
        return creatorId;
    }

    public void setCreatorId(String creatorId) {
        this.creatorId = creatorId;
    }

    public double getLatitude() {
        return latitude;
    }

    public void setLatitude(double latitude) {
        this.latitude = latitude;
    }

    public double getLongitude() {
        return longitude;
    }

    public void setLongitude(double longitude) {
        this.longitude = longitude;
    }

    public long getCreatedAtUTC() {
        return createdAtUTC;
    }

    public void setCreatedAtUTC(long createdAtUTC) {
        this.createdAtUTC = createdAtUTC;
    }
}

File: functions/src/models/TerrainAddress.ts
/* eslint-disable */

import {BadModelFieldException} from "../exceptions/BadModelFieldException";

export class TerrainAddress {

    private id!: string;
    private addressTitle!: string;
    private address!: string;
    private sport!: string;
    private priority!:number;
    private creatorId!: string;
    private latitude!: number;
    private longitude!: number;
    private createdAtUTC!: number;


    constructor() {
    }

    static makeInstanceFromSnapshot(snapshotNeedsParsing: any): TerrainAddress {

        const terrainAddress:TerrainAddress = new TerrainAddress();

        let parsedSnapshot;
        try{
            parsedSnapshot = JSON.parse(snapshotNeedsParsing);
        }catch (Error){
            parsedSnapshot = snapshotNeedsParsing;
        }

        console.log(parsedSnapshot);
        console.log(parsedSnapshot.id);
        terrainAddress.setId(parsedSnapshot.id);
        terrainAddress.setAddressTitle(parsedSnapshot.addressTitle);
        terrainAddress.setAddress(parsedSnapshot.address);
        terrainAddress.setSport(parsedSnapshot.sport);
        terrainAddress.setPriority(parsedSnapshot.priority);
        terrainAddress.setCreatorId(parsedSnapshot.creatorId);
        terrainAddress.setLatitude(parsedSnapshot.latitude);
        terrainAddress.setLongitude(parsedSnapshot.longitude);
        terrainAddress.setCreatedAtUTC(parsedSnapshot.createdAtUTC);

        return terrainAddress;
    }

    static makeCopy(terrainAddressToCopy: TerrainAddress | null) {

        if (terrainAddressToCopy === null)
            return null;

        const terrainAddress:TerrainAddress = new TerrainAddress();


        terrainAddress.setId(terrainAddressToCopy.getId());
        terrainAddress.setAddressTitle(terrainAddressToCopy.getAddressTitle());
        terrainAddress.setAddress(terrainAddressToCopy.getAddress());
        terrainAddress.setSport(terrainAddressToCopy.getSport());
        terrainAddress.setPriority(terrainAddressToCopy.getPriority());
        terrainAddress.setCreatorId(terrainAddressToCopy.getCreatorId());
        terrainAddress.setLatitude(terrainAddressToCopy.getLatitude());
        terrainAddress.setLongitude(terrainAddressToCopy.getLongitude());
        terrainAddress.setCreatedAtUTC(terrainAddressToCopy.getCreatedAtUTC());

        return terrainAddress;
    }


    toObject() {//this is the output that goes to firestore so becareful with the changes in this class


        return {
            id: this.id,
            addressTitle: this.addressTitle,
            address: this.address,
            sport: this.sport,
            priority: this.priority,
            creatorId: this.creatorId,
            latitude: this.latitude,
            longitude: this.longitude,
            createdAtUTC: this.createdAtUTC,
            // add other fields as needed
        };
    }

    public getId(): string {
        return this.id;
    }

    public setId(id: string) {
        if (id === undefined || typeof id !== 'string') {
            throw new BadModelFieldException("id must be a string");
        }

        id = id.trim();
        if (id.length == 0)
            throw new BadModelFieldException("id length is 0");


        let idLength:number = 60;
        if (id.length > idLength){
            throw new BadModelFieldException("id must be less than " + idLength + " letters");
        }

        this.id = id;
    }

    public getAddressTitle(): string {
        return this.addressTitle;
    }

    public setAddressTitle(addressTitle: string) {
        if (addressTitle === undefined || typeof addressTitle !== 'string') {
            throw new BadModelFieldException("address title must be a valid string");
        }

        let addressTitleMaxLength:number = 50;
        if (addressTitle.length > addressTitleMaxLength){
            throw new BadModelFieldException("address title max length must be less than " + addressTitleMaxLength + " letters");
        }

        this.addressTitle = addressTitle;
    }

    public getAddress(): string {
        return this.address;
    }

    public setAddress(address: string) {

        if (address === undefined || typeof address !== 'string') {
            throw new BadModelFieldException("address must be a valid string");
        }

        let addressMaxLength:number = 100;
        if (address.length > addressMaxLength){
            throw new BadModelFieldException("address max length must be less than " + addressMaxLength + " letters");
        }

        this.address = address;
    }

    public getSport(): string {
        return this.sport;
    }

    public setSport(sport: string) {

        if (sport === undefined || typeof sport !== 'string') {
            throw new BadModelFieldException("sport must be a valid string");
        }

        let sportMaxLength:number = 50;
        if (sport.length > sportMaxLength){
            throw new BadModelFieldException("sport max length must be less than " + sportMaxLength + " letters");
        }

        this.sport = sport;
    }

    public getPriority(): number {
        return this.priority;
    }

    public setPriority(priority: number) {

        if (priority === undefined || typeof priority !== 'number' || isNaN(priority)) {
            throw new BadModelFieldException("latitude must be a valid number");
        }

        const min:number = -2;
        if (priority <= min){
            throw new BadModelFieldException("priority must be more than " + min + "");
        }

        this.priority = priority;
    }

    public getCreatorId(): string {
        return this.creatorId;
    }

    public setCreatorId(creatorId: string) {

        if (creatorId === undefined || typeof creatorId !== 'string') {
            throw new BadModelFieldException("creator id must be a string");
        }

        let creatorIdLength:number = 60;
        if (creatorId.length > creatorIdLength){
            throw new BadModelFieldException("creator id must be less than " + creatorIdLength + " letters");
        }

        this.creatorId = creatorId;
    }

    public getLatitude(): number {
        return this.latitude;
    }

    public setLatitude(latitude: number) {

        if (latitude === undefined || typeof latitude !== 'number' || isNaN(latitude)) {
            throw new BadModelFieldException("latitude must be a valid number");
        }

        if (latitude < -90 || latitude > 90) {
            throw new BadModelFieldException("Invalid latitude. Latitude must be between -90 and 90 degrees.");
        }

        this.latitude = latitude;
    }

    public getLongitude(): number {
        return this.longitude;
    }

    public setLongitude(longitude: number) {

        if (longitude === undefined || typeof longitude !== 'number' || isNaN(longitude)) {
            throw new BadModelFieldException("longitude must be a valid number");
        }

        if (longitude < -180 || longitude > 180) {
            throw new BadModelFieldException("Invalid longitude. Longitude must be between -180 and 180 degrees.");
        }

        this.longitude = longitude;
    }

    public getCreatedAtUTC(): number {
        return this.createdAtUTC;
    }

    public setCreatedAtUTC(createdAtUTC: number) {

        if (createdAtUTC === undefined || typeof createdAtUTC !== 'number' || isNaN(createdAtUTC)) {
            throw new BadModelFieldException("createdAtUTC must be a valid number");
        }

        let minCreatedAtUTC:number = 0;
        if (createdAtUTC < minCreatedAtUTC){
            throw new BadModelFieldException("createdAtUTC shouldn't be less than " + minCreatedAtUTC);
        }

        this.createdAtUTC = createdAtUTC;
    }



}
